src/register_allocator.cc

```diff
@@ -0,0 +1,107 @@
+#include "register_allocator.h"
+
+#include <cassert>
+
+RegisterAllocator::RegisterAllocator(uint16_t num_physical_registers)
+{
+  for (uint16_t i = 0; i < num_physical_registers; ++i) {
+    free_registers.push(static_cast<PHYSICAL_REGISTER_ID>(i));
+  }
+  physical_register_file = std::vector<physical_register>(num_physical_registers, {0, 0, false, false});
+  frontend_RAT.fill(-1); //default value for no mapping
+  backend_RAT.fill(-1);
+}
+
+PHYSICAL_REGISTER_ID RegisterAllocator::rename_dest_register(int16_t reg, ooo_model_instr &instr)
+{
+  assert(free_registers.size() > 0);
+
+  PHYSICAL_REGISTER_ID phys_reg = free_registers.front();
+  free_registers.pop();
+  frontend_RAT[reg] = phys_reg;
+  physical_register_file[phys_reg] = {(uint16_t)reg, instr.instr_id, false, true}; //arch_reg_index, valid, busy
+
+  return phys_reg;
+}
+
+PHYSICAL_REGISTER_ID RegisterAllocator::rename_src_register(int16_t reg)
+{
+  PHYSICAL_REGISTER_ID phys = frontend_RAT[reg];
+
+  if (phys < 0) {
+    // allocate the register if it hasn't yet been mapped
+    // (common due to the traces being slices in the middle of a program)
+    phys = free_registers.front();
+    free_registers.pop();
+    frontend_RAT[reg] = phys;
+    physical_register_file[phys] = {(uint16_t)reg, 0, true, true}; //arch_reg_index, producing_inst_id, valid, busy
+  }
+
+  return phys;
+}
+
+void RegisterAllocator::complete_dest_register(PHYSICAL_REGISTER_ID physreg)
+{
+  // mark the physical register as valid
+  physical_register_file[physreg].valid = true;
+}
+
+void RegisterAllocator::retire_dest_register(PHYSICAL_REGISTER_ID physreg)
+{
+  // grab the arch reg index, find old phys reg in backend RAT
+  uint16_t arch_reg = physical_register_file[physreg].arch_reg_index;
+  PHYSICAL_REGISTER_ID old_phys_reg = backend_RAT[arch_reg];
+
+  // update the backend RAT with the new phys reg
+  backend_RAT[arch_reg] = physreg;
+
+  // free the old phys reg
+  if(old_phys_reg != -1) {
+    free_register(old_phys_reg);
+  }
+}
+
+void RegisterAllocator::free_register(PHYSICAL_REGISTER_ID physreg)
+{
+    physical_register_file[physreg] = {255, 0, false, false}; //arch_reg_index, producing_inst_id, valid, busy
+    free_registers.push(physreg);
+}
+
+bool RegisterAllocator::isValid(PHYSICAL_REGISTER_ID physreg) { return physical_register_file[physreg].valid; }
+
+unsigned long RegisterAllocator::count_free_registers() { return std::size(free_registers); }
+
+int RegisterAllocator::count_reg_dependencies(ooo_model_instr& instr) {
+  // mainly for unit tests. This is not used in the core model
+  int depcount = 0;
+  for (auto reg : instr.source_registers){
+    if (!isValid(reg)) depcount++;
+  }
+  return depcount;
+}
```

Comment:
Should this just be a free function in the test file?

Reply:
I had just finished moving it and updating all the calls when I realized I write this loop verbatim in the deadlock print, so it will be called there to avoid code repetition.

Reply:
```suggestion
int RegisterAllocator::count_reg_dependencies(ooo_model_instr& instr) {
  return std::count_if(std::begin(instr.source_registers), std::end(instr.source_registers), [](auto reg){ return !isValid(reg); });
}
```

Alright then. How about this instead?

Reply:
Looks good. Modern even.